[bug] get_cpus is not supporting cgroup2 #11635

MariaMozgunova opened the pull request:
Changelog: Fix: Get the cpu count for cgroup2 from `/sys/fs/cgroup/cpu.max`.
Docs: https://github.com/conan-io/docs/pull/2658

Fixes #11635

Gets the cpu count for cgroup2 from `/sys/fs/cgroup/cpu.max` file. From [docs](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html), `/sys/fs/cgroup/cpu.max` file has the following format: `$MAX $PERIOD`.
In case `$MAX` is "max", return 1.
If there is no `$PERIOD`, use only `$MAX` as the return value.
When both values are integers, use them as before.

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

Commit:
[bug] get_cpus is not supporting cgroup2 #11635

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=11667) <br/>All committers have signed the CLA.

MariaMozgunova commented:
Should I document the changes to the Conan docs?

memsharded commented:
> Should I document the changes to the Conan docs?

Better wait until it is merged, and then yes, a PR to the docs, if necessary, would be nice, just a couple of lines if it makes sense

MariaMozgunova commented:
Do I need to do something for this PR to be merged?

memsharded commented:
> Do I need to do something for this PR to be merged?

No, it is good, it just takes some time to review, too many ongoing things. It doesn't have tests, but honestly all tests are going to be pure mocks, so probably it is good.

Maybe @maikelvdh want to comment something more, if not, this will be merged and released in 1.51 (still a few days, Conan 2.0-beta.2 is in the next release)

maikelvdh commented on conan/tools/build/cpu.py:
```diff
@@ -15,8 +16,18 @@ def _cpu_count():
+            if os.path.exists("/sys/fs/cgroup/cgroup.controllers"):
+                cpu_max = load("/sys/fs/cgroup/cpu.max").split()
+                if cpu_max[0] == "max":
+                    return 1
```
In case of `max` it means basically no limits are established and therefore we should be able to rely on the `multiprocessing.cpu_count()`. In the current logic you would otherwise effectively in case of docker when cgroup V2 present and without any limiting of CPU cores applied just run with 1 jobs always.

memsharded replied:
But is it not an issue that ``multiprocessing.cpu_count()`` is used? because it will not match the docker assigned cores, and it can also overflow the container capacity?

maikelvdh replied:
That is exactly what you want in case of docker if you are not using the runtime options to limit the CPU (https://docs.docker.com/config/containers/resource_constraints/#cpu). In case you are using this options to limit you will get not `max` as value.

memsharded replied:
Great, thanks very much for the clarification. @MariaMozgunova could you please apply this in your PR? Thanks!

maikelvdh commented on conan/tools/build/cpu.py:
```diff
@@ -15,8 +16,18 @@ def _cpu_count():
+                if cpu_max[0] == "max":
+                    return 1
+                elif len(cpu_max) == 1:
+                    cfs_quota_us, cfs_period_us = int(cpu_max[0]), 1
```
To keep it in line with the actual official documentation and the microsecond notation
```suggestion
                    cfs_quota_us, cfs_period_us = int(cpu_max[0]), 100000
```

maikelvdh commented:
Thanks a lot @MariaMozgunova for addressing this problem 🚀 !

Commit:
Fix cgroup2 "max" case #11635

When $MAX == "max", there is no limit on the cpu usage set
Use multiprocessing.cpu_count() function in that case

memsharded commented:
Merged, thanks for your contribution! It will be part of 1.51

If you now want to do a PR to the docs, something very simple about the behavior for cgroups2, a couple of lines would be enough, that would be great.

MariaMozgunova commented:
@memsharded Can you please hint to me what file precisely should be modified?

uilianries commented:
@MariaMozgunova The documentation already describes cgroup usage, so with your hotfix, the follow part should updated:

https://github.com/conan-io/docs/blob/79feb32d0be4d531a363b8948b6d8a2289e059ab/reference/tools.rst#toolscpu_count

The website will look like this: https://docs.conan.io/en/latest/reference/tools.html#tools-cpu-count

Thank you for your contribution!
